File: MyApplication/app/src/main/java/com/bansi/myapplication/MainActivity.java
package com.bansi.myapplication;

import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity {
    private UserViewModel viewModel;
    private UserAdapter adapter;
    private boolean isLoading = false;
    private int page = 0;
    private final int PAGE_SIZE = 10;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new UserAdapter();
        recyclerView.setAdapter(adapter);

        viewModel = new ViewModelProvider(this).get(UserViewModel.class);
        loadUsers();

        viewModel.getUsersSortedByName().observe(this, adapter::setUsers);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                int visibleItemCount = layoutManager.getChildCount();
                int totalItemCount = layoutManager.getItemCount();
                int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

                if (!isLoading && (visibleItemCount + firstVisibleItemPosition >= totalItemCount) &&
                        firstVisibleItemPosition >= 0 && totalItemCount >= PAGE_SIZE) {
                    loadUsers();
                }
            }
        });
    }

    private void loadUsers() {
        isLoading = true;
        Toast.makeText(this, "page Load" +page, Toast.LENGTH_SHORT).show();
        viewModel.fetchUsers(PAGE_SIZE, page * PAGE_SIZE);
        page++;
        isLoading = false;
    }
}

File: MyApplication2/app/src/main/java/com/example/recipepaginationapp/ui/MainActivity.kt
package com.example.recipepaginationapp.ui

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import androidx.paging.LoadState
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bansi.myapplication.R
import com.example.recipepaginationapp.api.RecipeApiService
import com.example.recipepaginationapp.api.RetrofitClient
import com.example.recipepaginationapp.repository.RecipeRepository
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
class MainActivity : AppCompatActivity() {

    private lateinit var adapter: RecipeAdapter
    private lateinit var viewModel: RecipeViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main1)

        val apiService = RetrofitClient.instance.create(RecipeApiService::class.java)
        val repository = RecipeRepository(apiService)
        viewModel = RecipeViewModel(repository)

        adapter = RecipeAdapter { recipe ->
            val intent = Intent(this, FullScreenActivity::class.java)
            intent.putExtra("recipeList", ArrayList(adapter.snapshot().items))
            intent.putExtra("selectedRecipeId", recipe.id)
            startActivity(intent)
        }

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        recyclerView.layoutManager = GridLayoutManager(this,2)
        recyclerView.adapter = adapter

        lifecycleScope.launch {
            viewModel.recipes.collectLatest { pagingData ->
                adapter.submitData(pagingData)
            }
        }
    }
}

